source/particle/interpolator/bilinear_least_squares.cc

```diff
@@ -95,93 +101,172 @@ namespace aspect
+
+                    double max_total_slope = std::min(c[property_index][0] - property_bounds[property_index].first,
+                                                      property_bounds[property_index].second - c[property_index][0]);
+                    double current_total_slope = 0.0;
+                    for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                      {
+                        // The slope in any one direction should not overshoot/undershoot on its own.
+                        c[property_index][i] = std::copysign(std::min(std::abs(c[property_index][i]) * 2, max_total_slope), c[property_index][i]);
+                        current_total_slope += std::abs(c[property_index][i]);
+                      }
+                    current_total_slope *= half_h;
+                    if (current_total_slope > max_total_slope)
+                      {
+                        current_total_slope = 0;
+                        for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                          {
+                            c[property_index][i] = std::copysign(std::min(std::abs(c[property_index][i]), max_total_slope / half_h), c[property_index][i]);
+                            current_total_slope += std::abs(c[property_index][i]);
+                          }
+                        current_total_slope *= half_h;
+                        if (current_total_slope > max_total_slope)
+                          {
+                            double change_in_slope = (current_total_slope - max_total_slope)/(dim * half_h);
+                            dealii::Vector<double> c_i(n_matrix_columns);
+                            c_i[0] = c[property_index][0];
+                            for (unsigned int i = 1; i < n_matrix_columns; ++i)
+                              {
+                                c_i[i] = std::copysign(std::abs(c[property_index][i]) - change_in_slope, c[property_index][i]);
+                              }
+                            if (dim == 2)
+                              {
+                                c[property_index] = c_i;
+                              }
+                            else
+                              {
+                                // In two dimensions half the neccessary change can be safely removed from each slope component
+                                // however in three dimensions this additional check is neccessary to ensure that a third of the neccesary slope can be removed from each component.
+                                // An example where this could happen is if c is [.5, 1, .04, .01]' with boundaries of [0, 1].
+                                // Without these checks, c_3 would become negative, and would result in a different corner of the cell over(under)shooting.
+                                if (c[property_index][1] * c_i[1] <= 0)
```

Comment:
One question to also discuss when we meet to talk about this PR (just a note to myself to not forget about this thought): Is it possible that more than one component flip their sign? At the moment you only check if one of them changes.